File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cs
using System;
using System.Collections.Generic;

public class Solution {
    private char[] chars;
    private bool[] used;
    private HashSet<int> made;
    private List<char> cur;
    public int solution(string numbers) {
        
        chars = numbers.ToCharArray();
        used = new bool[chars.Length];
        made = new HashSet<int>();
        cur = new List<char>();
        
        DFS();
        int answer = 0;
        foreach(int x in made)
        {
            if(IsPrime(x)) answer++;
        }
        return answer;
    }
    private void DFS()
    {
        if(cur.Count > 0)
        {
            int val = int.Parse(new string(cur.ToArray()));
            made.Add(val);
        }
        if(cur.Count == chars.Length) return;
        
        for(int i = 0; i < chars.Length; i++)
        {
            if(!used[i])
            {
                used[i] = true;
                cur.Add(chars[i]);
                DFS();
                cur.RemoveAt(cur.Count - 1);
                used[i] = false;
            }
        }
    }
    
    private bool IsPrime(int n)
    {
        if(n <= 1) return false;
        if(n == 2) return true;
        if(n % 2 == 0) return false;
        int r = (int)Math.Sqrt(n);
        for(int i = 3; i <= r; i += 2)
        {
            if(n % i == 0) return false;
        }
        return true;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

set<int> s;
bool visited[10];
bool check(int n)
{
    if(n == 1) return false;
    if(n == 2) return true;
    if(n % 2 == 0) return false;
    for(int i = 3; i * i <= n; i++)
    {
        if(n % i == 0) return false;
    }
    return true;
}
void dfs(string numbers, string cur)
{
    if(cur != "")
        s.insert(stoi(cur));
    
    for(int i = 0; i < numbers.size(); i++)
    {
        if(visited[i]) continue;
        
        visited[i] = true;
        dfs(numbers, cur + numbers[i]);
        visited[i] = false;
    }
}

int solution(string numbers) {
    int answer = 0;

    dfs(numbers, "");
    
    for(int n : s)
    {
        if(check(n)) answer++;
    }
    return answer;
}